pkg/nsx/services/securitypolicy/builder.go

```diff
@@ -598,22 +580,43 @@ func (service *SecurityPolicyService) buildRuleOutGroup(obj *v1alpha1.SecurityPo
 	return nsxRuleDstGroup, nsxRuleSrcGroupPath, nsxRuleDstGroupPath, nsxGroupShare, nil
 }
 
-func (service *SecurityPolicyService) buildRuleID(obj *v1alpha1.SecurityPolicy, rule *v1alpha1.SecurityPolicyRule, ruleIdx int, createdFor string) string {
-	serializedBytes, _ := json.Marshal(rule)
-	ruleHash := fmt.Sprintf("%s", util.Sha1(string(serializedBytes)))
-	ruleIdxStr := fmt.Sprintf("%d", ruleIdx)
+func (service *SecurityPolicyService) buildRuleID(obj *v1alpha1.SecurityPolicy, ruleIdx int, createdFor string) string {
+	ruleHash := service.buildRuleHashString(&(obj.Spec.Rules[ruleIdx]))
+
 	if IsVPCEnabled(service) {
-		suffix := strings.Join([]string{ruleIdxStr, ruleHash}, common.ConnectorUnderline)
-		return util.GenerateIDByObjectWithSuffix(obj, suffix)
+		return util.GenerateIDByObjectWithSuffix(obj, ruleHash)
 	}
+
 	prefix := common.SecurityPolicyPrefix
 	if createdFor == common.ResourceTypeNetworkPolicy {
 		prefix = common.NetworkPolicyPrefix
 	}
+	ruleIdxStr := fmt.Sprintf("%d", ruleIdx)
 	return util.GenerateID(fmt.Sprintf("%s", obj.UID), prefix, ruleHash, ruleIdxStr)
 }
 
-func (service *SecurityPolicyService) buildRuleDisplayName(rule *v1alpha1.SecurityPolicyRule, portIdx, portNumber int, hasNamedport bool, createdFor string) (string, error) {
+// A rule containing named port may be expanded to multiple NSX rules if the name ports map to multiple port numbers.
+// So, in VPC network, the rule port numbers, which either are defined in rule Port or resolved from named port, will be appended as CR rule baseID to distinguish them.
+// For T1, the portIdx and portAddressIdx are appended as suffix.
+func (service *SecurityPolicyService) buildExpandedRuleID(obj *v1alpha1.SecurityPolicy, rule *v1alpha1.SecurityPolicyRule, ruleIdx int,
```

Comment:
Honestly, I don't like the flag `hasNamedPort` when building Rule's ID or display name, because it has no valid meaning in the final string. Can we make it more simple by directly telling the function what is the final port number or port range used in the output?